File: test_invitation_flow_full.js
const { createClient } = require('@supabase/supabase-js');

const SUPABASE_URL = "https://klyqeyuzugvabivfqlyt.supabase.co";
const SUPABASE_ANON_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImtseXFleXV6dWd2YWJpdmZxbHl0Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NDAxNjg0MjUsImV4cCI6MjA1NTc0NDQyNX0.FhM9c8tFpP6eFwK2Jb_tC1JvX7M9-n3yG3L9l1m3p4Q";

const clientA = createClient(SUPABASE_URL, SUPABASE_ANON_KEY);
const clientB = createClient(SUPABASE_URL, SUPABASE_ANON_KEY);
const clientC = createClient(SUPABASE_URL, SUPABASE_ANON_KEY);

async function runInvitationFlowTest() {
    console.log("🚀 STARTING REAL SUPABASE TEST FOR MATCHING -> SELECTION -> INVITATION FLOW...");

    // 1. Authenticate User A, B, C
    const emailA = "req_user_a@lyann.app";
    const emailB = "req_user_b@lyann.app";
    const emailC = "req_user_c@lyann.app";
    const password = "Password123!";

    async function getOrCreateUser(client, email) {
        let authRes = await client.auth.signInWithPassword({ email, password });
        if (authRes.error) {
            authRes = await client.auth.signUp({
                email,
                password,
                options: { data: { first_name: email.split('@')[0], last_name: 'Test' } }
            });
            if (authRes.error) throw authRes.error;
        }
        return authRes.data.user;
    }

    const userA = await getOrCreateUser(clientA, emailA);
    const userB = await getOrCreateUser(clientB, emailB);
    const userC = await getOrCreateUser(clientC, emailC);

    console.log(`✅ Authenticated User A (${userA.id}), User B (${userB.id}), User C (${userC.id})`);

    // Ensure profiles exist in DB for matching
    await clientA.from('profiles').upsert({ id: userA.id, email: emailA, first_name: 'UserA', city: 'Le Gosier' });
    await clientB.from('profiles').upsert({ id: userB.id, email: emailB, first_name: 'UserB', city: 'Le Gosier' });
    await clientC.from('profiles').upsert({ id: userC.id, email: emailC, first_name: 'UserC', city: 'Les Abymes' });

    const createdRequestIds = [];
    try {
        // 2. User A publishes Request 1 (Jardinage / Débroussaillage)
        const { data: req1, error: req1Err } = await clientA
            .from('requests')
            .insert({
                requester_id: userA.id,
                title: 'Jardinage - Débroussaillage',
                description: 'Besoin d\'un coup de main pour débroussailler mon jardin',
                category: 'Jardinage',
                location: 'Le Gosier',
                budget: 50,
                urgency: 'Flexible',
                status: 'OPEN'
            })
            .select()
            .single();

        if (req1Err) throw new Error("Request 1 creation failed: " + req1Err.message);
        createdRequestIds.push(req1.id);
        console.log(`✅ User A created Request 1: ID ${req1.id}`);

        // 3. User A selects ONLY User B (C is not selected)
        // Send invitation to B
        const { data: sendRes, error: sendErr } = await clientA.rpc('send_request_invitations', {
            p_request_id: req1.id,
            p_recipient_ids: [userB.id]
        });

        if (sendErr) throw new Error("send_request_invitations failed: " + sendErr.message);
        console.log("✅ send_request_invitations result:", sendRes);

        // 4. Verify received invitations for B and C
        const { data: bInvs, error: bInvsErr } = await clientB
            .from('request_invitations')
            .select('*')
            .eq('recipient_id', userB.id)
            .eq('request_id', req1.id);

        if (bInvsErr) throw bInvsErr;
        console.log(`✅ User B received ${bInvs.length} invitation(s) for Request 1. Status: ${bInvs[0]?.status}`);
        if (bInvs.length !== 1 || bInvs[0].status !== 'PENDING') {
            throw new Error("FAIL: User B should have received 1 PENDING invitation!");
        }

        const { data: cInvs } = await clientC
            .from('request_invitations')
            .select('*')
            .eq('recipient_id', userC.id)
            .eq('request_id', req1.id);

        console.log(`✅ User C received ${cInvs ? cInvs.length : 0} invitation(s) for Request 1.`);
        if (cInvs && cInvs.length > 0) {
            throw new Error("FAIL: User C was not selected but received an invitation!");
        }

        // 5. User B accepts the invitation
        const invIdB = bInvs[0].id;
        const { data: acceptRes, error: acceptErr } = await clientB.rpc('accept_request_invitation', {
            p_invitation_id: invIdB
        });

        if (acceptErr) throw new Error("accept_request_invitation failed: " + acceptErr.message);
        console.log("✅ User B accepted invitation result:", acceptRes);

        if (acceptRes.conversation_id && acceptRes.request_id === req1.id) {
            console.log(`✅ Verified: Invitation status ACCEPTED, conversation_id = ${acceptRes.conversation_id}, request_id = ${acceptRes.request_id}`);
        } else {
            throw new Error("FAIL: Invalid accept_request_invitation return structure!");
        }

        // 6. Security tests:
        // a) User C attempts to accept User B's invitation -> Should fail
        const { error: cAcceptErr } = await clientC.rpc('accept_request_invitation', {
            p_invitation_id: invIdB
        });
        console.log("🔒 User C accept attempt result (expected error):", cAcceptErr ? cAcceptErr.message : "NO ERROR (UNEXPECTED)");
        if (!cAcceptErr) {
            throw new Error("FAIL: User C was able to accept B's invitation!");
        }

        // b) User A attempts to invite self (A -> A) -> Should fail or return 0 inserted
        const { data: selfRes } = await clientA.rpc('send_request_invitations', {
            p_request_id: req1.id,
            p_recipient_ids: [userA.id]
        });
        console.log("🔒 Self-invitation attempt result:", selfRes);
        if (selfRes.inserted_count !== 0) {
            throw new Error("FAIL: User A was able to invite self!");
        }

        // c) User A selects B twice in array -> Should only create 1 invitation row
        const { data: dupRes } = await clientA.rpc('send_request_invitations', {
            p_request_id: req1.id,
            p_recipient_ids: [userB.id, userB.id]
        });
        console.log("🔒 Duplicate selection result:", dupRes);

        // 7. Multi-request context test (Requirement 8):
        // User A creates Request 2 (Plomberie) and invites B
        const { data: req2, error: req2Err } = await clientA
            .from('requests')
            .insert({
                requester_id: userA.id,
                title: 'Plomberie - Réparation fuite',
                description: 'Fuite sous evier cuisine',
                category: 'Plomberie',
                location: 'Le Gosier',
                budget: 70,
                urgency: 'Urgent',
                status: 'OPEN'
            })
            .select()
            .single();

        if (req2Err) throw new Error("Request 2 creation failed: " + req2Err.message);
        createdRequestIds.push(req2.id);
        console.log(`✅ User A created Request 2: ID ${req2.id}`);

        await clientA.rpc('send_request_invitations', {
            p_request_id: req2.id,
            p_recipient_ids: [userB.id]
        });

        const { data: bReq2Invs } = await clientB
            .from('request_invitations')
            .select('*')
            .eq('recipient_id', userB.id)
            .eq('request_id', req2.id);

        const invIdReq2 = bReq2Invs[0].id;
        const { data: acceptRes2 } = await clientB.rpc('accept_request_invitation', {
            p_invitation_id: invIdReq2
        });

        console.log("✅ User B accepted Request 2 invitation result:", acceptRes2);

        // Verify distinct request contexts between same users A and B
        const { data: allABInvs } = await clientB
            .from('request_invitations')
            .select('id, request_id, requester_id, recipient_id, conversation_id, status')
            .eq('requester_id', userA.id)
            .eq('recipient_id', userB.id);

        console.log("✅ All invitations between A and B:", allABInvs);

        if (allABInvs.length >= 2) {
            const reqIds = allABInvs.map(i => i.request_id);
            const uniqueReqIds = new Set(reqIds);
            if (uniqueReqIds.size >= 2) {
                console.log("🎉 VERIFIED: Distinct business contexts (request_ids) preserved between same users A and B!");
            } else {
                throw new Error("FAIL: Lost request_id context!");
            }
        } else {
            throw new Error("FAIL: Expected at least 2 invitations between A and B!");
        }

        console.log("\n==========================================");
        console.log("SUCCESS: ALL MATCHING -> SELECTION -> INVITATION TESTS PASSED!");
        console.log("==========================================\n");
    } finally {
        if (createdRequestIds.length > 0) {
            console.log("🧹 [E2E Teardown] Cleaning up created test requests:", createdRequestIds);
            for (const rId of createdRequestIds) {
                try {
                    await clientA.from('requests').delete().eq('id', rId);
                } catch(e) {
                    console.warn("Cleanup error for request:", rId, e);
                }
            }
        }
    }
}

runInvitationFlowTest().catch(err => {
    console.error("❌ TEST FAILED:", err);
    process.exit(1);
});

File: test_invitation_flow_full.py
import json
import urllib.request
import urllib.error
import ssl
import sys

SUPABASE_URL = "https://gzispjfoywklpqatjyop.supabase.co"
SUPABASE_ANON_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imd6aXNwamZveXdrbHBxYXRqeW9wIiwicm9sZSI6ImFub24iLCJpYXQiOjE3ODY0MTE4ODcsImV4cCI6MjEwMTk4Nzg4N30.oPJvkDVffQ4EaxDy2V7Jj7veusCVGTnM2BUBKXnoQ0A"

headers_base = {
    "apikey": SUPABASE_ANON_KEY,
    "Content-Type": "application/json",
    "Prefer": "return=representation"
}

ssl_context = ssl._create_unverified_context()

def http_req(url, method="GET", body=None, token=None):
    hdrs = dict(headers_base)
    if token:
        hdrs["Authorization"] = f"Bearer {token}"

    data = json.dumps(body).encode('utf-8') if body is not None else None
    req = urllib.request.Request(url, data=data, headers=hdrs, method=method)

    try:
        with urllib.request.urlopen(req, context=ssl_context) as resp:
            resp_body = resp.read().decode('utf-8')
            return json.loads(resp_body) if resp_body else None, resp.status
    except urllib.error.HTTPError as e:
        err_body = e.read().decode('utf-8')
        return json.loads(err_body) if err_body else {"error": str(e)}, e.code

def login_user(email, password="Password123!"):
    url_signin = f"{SUPABASE_URL}/auth/v1/token?grant_type=password"
    body = {"email": email, "password": password}
    res, status = http_req(url_signin, method="POST", body=body)

    if status == 200 and "access_token" in res:
        return res["access_token"], res["user"]["id"]

    raise Exception(f"Failed to authenticate {email}: {res}")

def main():
    print("🚀 STARTING REAL SUPABASE PYTHON TEST FOR MATCHING -> SELECTION -> INVITATION FLOW...")

    # 1. Authenticate A & B
    tokenA, idA = login_user("req_user_a@lyann.app")
    tokenB, idB = login_user("req_user_b@lyann.app")
    idC = "b5051e99-0e32-4c2e-a24d-adf7dc02f9e7" # User C (unselected recipient)

    print(f"✅ User A Authenticated: ID {idA}")
    print(f"✅ User B Authenticated: ID {idB}")
    print(f"✅ User C Registered: ID {idC}")

    created_request_ids = []
    try:
        # Upsert profiles for A, B, C
        url_prof = f"{SUPABASE_URL}/rest/v1/profiles"
        http_req(url_prof, method="POST", body={"id": idA, "email": "req_user_a@lyann.app", "first_name": "UserA", "city": "Le Gosier"}, token=tokenA)
        http_req(url_prof, method="POST", body={"id": idB, "email": "req_user_b@lyann.app", "first_name": "UserB", "city": "Le Gosier"}, token=tokenB)

        # 2. User A creates Request 1
        url_req = f"{SUPABASE_URL}/rest/v1/requests"
        body_req1 = {
            "requester_id": idA,
            "title": "Jardinage - Débroussaillage Le Gosier",
            "description": "Besoin d'un coup de main pour débroussailler",
            "category": "Jardinage",
            "location": "Le Gosier",
            "budget": 50,
            "urgency": "Flexible",
            "status": "OPEN"
        }
        res_req1, status = http_req(url_req, method="POST", body=body_req1, token=tokenA)
        req1_id = res_req1[0]["id"]
        created_request_ids.append(req1_id)
        print(f"✅ User A created Request 1: ID {req1_id}")

        # 3. User A sends invitation ONLY to User B
        url_rpc_send = f"{SUPABASE_URL}/rest/v1/rpc/send_request_invitations"
        body_send = {"p_request_id": req1_id, "p_recipient_ids": [idB]}
        res_send, status = http_req(url_rpc_send, method="POST", body=body_send, token=tokenA)
        print(f"✅ send_request_invitations response: {res_send} (status {status})")

        # 4. User B checks received invitations
        url_b_invs = f"{SUPABASE_URL}/rest/v1/request_invitations?recipient_id=eq.{idB}&request_id=eq.{req1_id}"
        b_invs, status = http_req(url_b_invs, method="GET", token=tokenB)
        print(f"✅ User B received {len(b_invs)} invitation(s). Status: {b_invs[0]['status']}")
        if len(b_invs) != 1 or b_invs[0]['status'] != 'PENDING':
            raise Exception("FAIL: User B should have 1 PENDING invitation!")

        inv_id_b = b_invs[0]['id']

        # 5. User C checks received invitations -> Expect 0
        url_c_invs = f"{SUPABASE_URL}/rest/v1/request_invitations?recipient_id=eq.{idC}&request_id=eq.{req1_id}"
        c_invs, status = http_req(url_c_invs, method="GET", token=tokenA)
        print(f"✅ User C received {len(c_invs)} invitation(s) for Request 1.")
        if len(c_invs) > 0:
            raise Exception("FAIL: User C was not selected but received invitation!")

        # 6. User B accepts the invitation
        url_rpc_accept = f"{SUPABASE_URL}/rest/v1/rpc/accept_request_invitation"
        body_accept = {"p_invitation_id": inv_id_b}
        res_accept, status = http_req(url_rpc_accept, method="POST", body=body_accept, token=tokenB)
        print(f"✅ User B accept_request_invitation response: {res_accept}")

        if not res_accept.get('success') or res_accept.get('request_id') != req1_id:
            raise Exception("FAIL: Accept invitation invalid response structure!")

        conv_id_1 = res_accept['conversation_id']
        print(f"✅ Verified: Invitation ACCEPTED, conversation_id = {conv_id_1}, request_id = {req1_id}")

        # 7. RLS / Security Checks:
        # a) Non-recipient (User A) attempts to accept B's invitation -> Should fail with permission error
        res_a_accept, status_a_accept = http_req(url_rpc_accept, method="POST", body={"p_invitation_id": inv_id_b}, token=tokenA)
        print(f"🔒 Non-recipient accept attempt status: {status_a_accept}, response: {res_a_accept}")
        if status_a_accept < 400 and isinstance(res_a_accept, dict) and res_a_accept.get('success') is True:
            raise Exception("FAIL: Unauthorized user accepted invitation!")

        # b) User A attempts to invite self (A -> A) -> Should return inserted_count 0
        res_self, status = http_req(url_rpc_send, method="POST", body={"p_request_id": req1_id, "p_recipient_ids": [idA]}, token=tokenA)
        print(f"🔒 Self-invitation attempt response: {res_self}")
        if res_self.get('inserted_count', 0) != 0:
            raise Exception("FAIL: User A invited self!")

        # c) Duplicate selection (A selects B twice) -> Should ignore duplicates
        res_dup, status = http_req(url_rpc_send, method="POST", body={"p_request_id": req1_id, "p_recipient_ids": [idB, idB]}, token=tokenA)
        print(f"🔒 Duplicate selection attempt response: {res_dup}")

        # 8. Multi-request context test (Requirement 8):
        body_req2 = {
            "requester_id": idA,
            "title": "Plomberie - Réparation fuite évier",
            "description": "Fuite sous évier cuisine",
            "category": "Plomberie",
            "location": "Le Gosier",
            "budget": 75,
            "urgency": "Urgent",
            "status": "OPEN"
        }
        res_req2, _ = http_req(url_req, method="POST", body=body_req2, token=tokenA)
        req2_id = res_req2[0]["id"]
        created_request_ids.append(req2_id)
        print(f"✅ User A created Request 2: ID {req2_id}")

        # Invite B for Request 2
        res_send2, status = http_req(url_rpc_send, method="POST", body={"p_request_id": req2_id, "p_recipient_ids": [idB]}, token=tokenA)
        print(f"✅ User A invited B for Request 2: {res_send2}")

        # B accepts Request 2
        b_invs2, _ = http_req(f"{SUPABASE_URL}/rest/v1/request_invitations?recipient_id=eq.{idB}&request_id=eq.{req2_id}", method="GET", token=tokenB)
        inv_id_b2 = b_invs2[0]['id']
        res_accept2, _ = http_req(url_rpc_accept, method="POST", body={"p_invitation_id": inv_id_b2}, token=tokenB)
        print(f"✅ User B accepted Request 2 invitation: {res_accept2}")

        # Verify both invitations exist with distinct request_id
        url_all_ab = f"{SUPABASE_URL}/rest/v1/request_invitations?requester_id=eq.{idA}&recipient_id=eq.{idB}"
        all_ab, _ = http_req(url_all_ab, method="GET", token=tokenB)
        print(f"✅ Total invitations between A and B in DB: {len(all_ab)}")
        req_ids_set = set(inv["request_id"] for inv in all_ab)
        print(f"✅ Unique request_ids stored in invitations: {req_ids_set}")

        if len(req_ids_set) >= 2:
            print("🎉 VERIFIED: Distinct request_id context maintained per invitation!")
        else:
            raise Exception("FAIL: lost distinct request_id context!")

        print("\n==========================================")
        print("ALL TESTS PASSED SUCCESSFULLY ON REAL SUPABASE PRODUCTION DB!")
        print("==========================================\n")
    finally:
        if created_request_ids:
            print("🧹 [Teardown] Cleaning up created test requests:", created_request_ids)
            for r_id in created_request_ids:
                try:
                    http_req(f"{SUPABASE_URL}/rest/v1/request_invitations?request_id=eq.{r_id}", method="DELETE", token=tokenA)
                    http_req(f"{SUPABASE_URL}/rest/v1/requests?id=eq.{r_id}", method="DELETE", token=tokenA)
                except Exception as e:
                    print(f"Teardown error for request {r_id}: {e}")

if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print(f"❌ TEST FAILED: {e}")
        sys.exit(1)
